### DISCUSSION TITLE
Micropython SPWM

### DISCUSSION THREAD

**Discussion by marestudio2004** [micropython/micropython#9342] [General]
Hello,

does anyone have experience with generating a sine wave using pwm (on ESP32 or 8622), need to drive MOSFETs for a single phase inverter like:

https://www.homemade-circuits.com/arduino-pure-sine-wave-inverter-circuit/

**Comment by jimmo** [comment]
Woah. Do not use this circuit to generate mains voltages! Buy a real inverter instead.

**Comment by robert-hh** [comment]
The circuit in the link does not generate a sine wave. It create for each half of the sine wave a 7 cycle PWM signal which emulates the energy of a sine wave. A 50Hz transformer is slow, so it will not carry over much of that 350Hz PWM, and the remaining will be reduced by the capacitor in the output. That circuit does not look overly dangerous, because all you touch & work with is at the low voltage side. Using a 3.3V device could be a problem, because it needs like 5V to get into a low Rds state and switch higher currents w/o taking too much power itself.

**Comment by marestudio2004** [comment]
I understand your concern, but I didn't get an answer to the given question... I assume that with PWM, it is necessary to make a dynamic change in the duty cycle parameter in a certain range... Is there a ready-made function or something like that... .. what happens when LED lighting changes color, can this function be used somehow?

**Reply by marestudio2004** [reply]
![image](https://user-images.githubusercontent.com/71499558/190877165-d62035ce-2f2a-40c7-9966-7fc2c586e4e7.png)

**Comment by robert-hh** [comment]
You might be able to use machine.bitstream() for your purpose, which has the bitstream ready for one half cycle and is repeated with alternating the drives every half cycle. Repeating would be started by the timer at a rate of 100 (or 120) Hz.
But with 800Hz pulse rate  and a timer it should be possible to make a software coded PWM, at least with the ESP32.  The ESP8266 with Python is probably too slow. The timer would be called at the PWM rate. Here it's 800 Hz, and it would then set a Pin, wait a certain time taken from a table and then reset the pin. The board would hardly be able to do something else. Using the time will ensure a stable frequency.

**Attention**: Your picture above is wrong. Only one of the output must be active at a certain time. So in the fist half of the cycle, you have only positive pulses, in the second only negative pulses. Otherwise it is not a push-pull operation, and

**Comment by robert-hh** [comment]
Forget bitstream. It does not work for that purpose. So you have to do a `manual` PWM.

**Comment by robert-hh** [comment]
I hacked together a short script. Times to be varied.
```
from machine import Pin, Timer, idle
import time
table = (100, 200, 400, 600, 800, 800, 600, 400, 200, 100)
count = 0
alt = 0
p9 = Pin(9, Pin.OUT, value=0)
p10 = Pin(10, Pin.OUT, value=0)

def pulse(t):
    global table, count, alt
    global p9, p10
    if alt:
        p9(1)
        time.sleep_us(table[count])
        p9(0)
    else:
        p10(1)
        time.sleep_us(table[count])
        p10(0)
    count = (count + 1)
    if count == len(table):
        alt = 1 - alt
        count = 0

t = Timer(callback=pulse, freq=1000, mode=Timer.PERIODIC)
while True:
    idle()
```
It creates the following waveform at Pins 9 and 10:
![alt_pulse](https://user-images.githubusercontent.com/12476868/190901953-d06d3521-25b5-49e8-8269-7fece7991de2.png)

**Reply by marestudio2004** [reply]
Thanks Mr. Rober! This is nice solution!

**Reply by marestudio2004** [reply]
This is table from C code, for DC/AC inverter SPWM based:

// choose 1 of the SPWM tables below:
const int sPWMArray[] = {500,500,750,500,1250,500,2000,500,1250,500,750,500,500}; // this is the array with the 50Hz SPWM values; change 

them at will

// const int sPWMArray[] = {416,416,625,416,1041,416,1666,416,1041,416,625,416,416,7}; // use this table to obtain 60Hz 

Values are given in us (microseconds)!

**Comment by robert-hh** [comment]
I had similar numbers initially, but changed them a little bit to make the difference more obvious in the diagram. The pulse lengths at the start and end must be longer.
Since the time can only be set up in periods of while ms, the slots per half wave are either 10 or 5.  In my example, I use 10, but 5 should be fine as well. At 10 slots, each slots has a length of 1000µs, and 5 slots 2000µs. The code cannot use the full slot length, because the callback must return before the next timer IRQ. Using the timer ensures a stable frequency over time. But you can as well just use a permanent loop and accept the drift.

**Comment by jimmo** [comment]
I'm going to re-iterate for @marestudio2004 and anyone else who might be reading this in the future: **This is a really bad idea. Inverters are cheap, just buy a real one.**

You risk damaging your devices, not to mention electrocuting yourself, or starting a fire. This circuit has no safety protections and is a hazard.
